lightning/src/routing/network_graph.rs

```diff
@@ -1040,6 +1063,67 @@ impl NetworkGraph {
 		}
 	}
 
+	#[cfg(feature = "std")]
+	/// Removes information about channels which we haven't heard any updates about in some time.
+	/// This can be used regularly to prune the network graph from channels which likely no longer
+	/// exist.
+	///
+	/// While there is no formal requirement that nodes regularly re-broadcast their channel
+	/// updates every two weeks, the non-normative section of BOLT 7 currently suggests that
+	/// pruning occurs for updates which are at least two weeks old, which we implement here.
+	///
+	/// This method is automatically called immediately before writing the network graph via
+	/// [`Writeable::write`].
+	///
+	/// This method is only available with the `std` feature. See
+	/// [`NetworkGraph::remove_stale_channels_with_time`] for `no-std` use.
+	pub fn remove_stale_channels(&self) {
+		let time = SystemTime::now().duration_since(UNIX_EPOCH).expect("Time must be > 1970").as_secs();
+		self.remove_stale_channels_with_time(time);
+	}
+
+	/// Removes information about channels which we haven't heard any updates about in some time.
+	/// This can be used regularly to prune the network graph from channels which likely no longer
+	/// exist.
+	///
+	/// While there is no formal requirement that nodes regularly re-broadcast their channel
+	/// updates every two weeks, the non-normative section of BOLT 7 currently suggests that
+	/// pruning occurrs for updates which are at least two weeks old, which we implement here.
+	///
+	/// This function takes the current unix time as an argument. For users with the `std` feature
+	/// enabled, [`NetworkGraph::remove_stale_channels`] may be preferable.
+	pub fn remove_stale_channels_with_time(&self, current_time_unix: u64) {
+		let mut channels = self.channels.write().unwrap();
+		// Time out if we haven't received an update in at least 14 days.
+		let min_time_unix: u32 = (current_time_unix - STALE_CHANNEL_UPDATE_AGE_LIMIT_SECS) as u32;
+		// Sadly BTreeMap::retain was only stabilized in 1.53 so we can't switch to it for some
+		// time.
+		let mut scids_to_remove = Vec::new();
+		for (scid, info) in channels.iter_mut() {
+			if info.one_to_two.is_some() && info.one_to_two.as_ref().unwrap().last_update < min_time_unix {
+				info.one_to_two = None;
+			}
+			if info.two_to_one.is_some() && info.two_to_one.as_ref().unwrap().last_update < min_time_unix {
+				info.two_to_one = None;
+			}
+			if info.one_to_two.is_none() && info.two_to_one.is_none() {
+				// We check the announcement_received_time here to ensure we don't drop
+				// announcements that we just received and are just waiting for our peer to send a
+				// channel_update for.
+				if info.announcement_received_time < min_time_unix as u64 {
+					scids_to_remove.push(*scid);
+				}
```

Comment:
BOLT 7 says using the oldest channel update rather than the time of the channel announcement. Likewise, the docs above more generally says we prune for updates older than two weeks.

Reply:
Right? I'm not sure what your're referring to here? We remove the directional info (disabling a direction) for each direction if the update is old, and, if both directions are gone, we remove the channel as long as the announcement was a while ago. I guess the docs could mention that we remove info about each direction individually, but I'm not sure it matters?

Reply:
Note that the channel-announcement receive-time check is unrelated to the pruning and exists only to avoid a race condition when downloading - if we're currently syncing the graph, its possible that we've received an announcement but not yet received updates for that channel. In that case, we'd remove the channel before receiving the updates, which we don't want to do.

Reply:
I see. I guess my reading of the BOLT says that once either update is older than two weeks, then the channel should be pruned. But here we wait for both.

Reply:
Yea, I guess we could make it either, no strong feelings, but note that the section on pruning is only in the discussion, not in the normative section. Thus, we don't need to start reading it in close detail - its not a specific thing for us to do.

Reply:
Fine with keeping this as is.